File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/organisms/header/OrgNavigationHeaderOverlay.kt
package com.perrystreet.woof.designsystem.atomic.organisms.header

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.WindowInsetsSides
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.only
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawing
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.button.AtomIconButton
import com.perrystreet.woof.designsystem.atomic.atoms.button.roles.ButtonBackgroundRole
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.organisms.overflowmenu.OrgOverflowMenuButton
import com.perrystreet.woof.designsystem.atomic.organisms.overflowmenu.OrgOverflowMenuItem
import com.perrystreet.woof.resources.R

@Composable
fun OrgNavigationHeaderOverlay(
    onBackTap: () -> Unit,
    actions: List<OrgNavigationHeaderActionItem> = emptyList(),
    overflowItems: List<OrgOverflowMenuItem> = emptyList(),
    isOverflowExpanded: Boolean = false,
    onOverflowExpandedChange: (Boolean) -> Unit = {},
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Top + WindowInsetsSides.Horizontal))
            .height(SizingRoles.InteractionHeight.Comfort.dp)
            .padding(horizontal = PaddingRoles.Element.Compact.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        AtomIconButton(
            iconRes = R.drawable.ic_arrow_back,
            contentDescription = stringResource(R.string.accessibility_back),
            colorRole = IconColorRole.OnScrim,
            backgroundRole = ButtonBackgroundRole.None,
            onTap = onBackTap,
        )
        Spacer(modifier = Modifier.weight(1f))
        actions.forEach { action ->
            AtomIconButton(
                iconRes = action.iconRes,
                contentDescription = action.contentDescription,
                colorRole = action.colorRole,
                backgroundRole = ButtonBackgroundRole.None,
                onTap = action.onTap,
            )
        }
        OrgOverflowMenuButton(
            items = overflowItems,
            isExpanded = isOverflowExpanded,
            onExpandedChange = onOverflowExpandedChange,
            colorRole = IconColorRole.OnScrim,
        )
    }
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Organisms/Header/OrgNavigationHeaderOverlay.swift
import Resources
import SwiftUI

public struct OrgNavigationHeaderOverlay: View {
    private let onBackTap: () -> Void
    private let actions: [OrgNavigationHeaderActionItem]
    private let overflowItems: [OrgOverflowMenuItem]
    private let isOverflowExpanded: Bool
    private let onOverflowExpandedChange: (Bool) -> Void

    public init(
        onBackTap: @escaping () -> Void,
        actions: [OrgNavigationHeaderActionItem] = [],
        overflowItems: [OrgOverflowMenuItem] = [],
        isOverflowExpanded: Bool = false,
        onOverflowExpandedChange: @escaping (Bool) -> Void = { _ in }
    ) {
        self.onBackTap = onBackTap
        self.actions = actions
        self.overflowItems = overflowItems
        self.isOverflowExpanded = isOverflowExpanded
        self.onOverflowExpandedChange = onOverflowExpandedChange
    }

    public var body: some View {
        HStack(spacing: 0) {
            AtomIconButton(
                icon: Asset.Icons.arrowBack,
                contentDescription: L10n.Accessibility.back,
                colorRole: .onScrim,
                backgroundRole: .none,
                onTap: onBackTap
            )
            Spacer(minLength: 0)
            ForEach(Array(actions.enumerated()), id: \.offset) { action in
                AtomIconButton(
                    icon: action.element.icon,
                    contentDescription: action.element.contentDescription,
                    colorRole: action.element.colorRole,
                    backgroundRole: .none,
                    onTap: action.element.onTap
                )
            }
            OrgOverflowMenuButton(
                items: overflowItems,
                isExpanded: isOverflowExpanded,
                onExpandedChange: onOverflowExpandedChange,
                colorRole: .onScrim
            )
        }
        .padding(.horizontal, PaddingRoles.Element.compact.rawValue)
        .frame(maxWidth: .infinity, minHeight: SizingRoles.InteractionHeight.comfort.rawValue)
    }
}
